aphrodite/modeling/layers/sampler.py

```diff
@@ -507,25 +508,39 @@ def _apply_penalties(logits: torch.Tensor, prompt_tokens_tensor: torch.Tensor,
                      output_tokens_tensor: torch.Tensor,
                      presence_penalties: torch.Tensor,
                      frequency_penalties: torch.Tensor,
-                     repetition_penalties: torch.Tensor) -> torch.Tensor:
+                     repetition_penalties: torch.Tensor,
+                     rep_range: Optional[int] = None) -> torch.Tensor:
+    """Apply presence, frequency, and repetition penalties to the logits."""
     num_seqs, vocab_size = logits.shape
-    _, prompt_mask = _get_bin_counts_and_mask(prompt_tokens_tensor, vocab_size,
-                                              num_seqs)
-    output_bin_counts, output_mask = _get_bin_counts_and_mask(
-        output_tokens_tensor, vocab_size, num_seqs)
+    
+    if rep_range is not None and rep_range > 0:
+        # Just take the last rep_range tokens from output_tokens_tensor
+        # This is much more efficient as we're only looking at recent history
+        output_tokens_tensor = output_tokens_tensor[:, -rep_range:]
```

Comment:
This is applying the same range to all sequences in batch, no?

Also creating a new tensor here is probably less efficient. I think the slicing should be done in the bin counting function above this.

Reply:
Agreed, I'll have to go over it again with batching in mind, and also agreed about rolling in the slicing